website/src/models/group.js

```diff
@@ -440,69 +440,56 @@ schema.statics.bossQuest = function bossQuest (user, progress) {
-      ).then(() => challenges); // pass `challenges` above to next promise TODO ok to return a non-promise?
-    }).then(challenges => {
-      return Q.all(challenges.map(chal => {
-        let i = user.challenges.indexOf(chal._id);
-        if (i !== -1) user.challenges.splice(i, 1);
-        return user.unlink({cid: chal._id, keep});
-      }));
-    }),
-
-    // Update the group
-    (() => {
-      // If user is the last one in group and group is private, delete it
-      if (group.members.length === 1 && (
-          group.type === 'party' ||
-          group.type === 'guild' && group.privacy === 'private'
-      )) return group.remove();
-
-      let update = {};
-      // otherwise just remove a member TODO create User.methods.removeFromGroup?
-      if (group.type === 'guild') {
-        _.pull(user.guilds, group._id);
-      } else {
-        user.party._id = undefined;
-      }
+  let challenges = await Challenge.find({
+    _id: {$in: user.challenges}, // Challenges I am in
+    group: group._id, // that belong to the group I am leaving
+  });
 
-      // If the leader is leaving (or if the leader previously left, and this wasn't accounted for)
-      let leader = group.leader;
+  await Challenge.update(
+    {_id: {$in: _.pluck(challenges, '_id')}},
+    {$pull: {members: user._id}},
+    {multi: true}
+  );
+
+  //  remove user from challenges
+  let challengesUserWasRemovedFrom = challenges.map(chal => {
+    let i = user.challenges.indexOf(chal._id);
+    if (i !== -1) user.challenges.splice(i, 1);
+    return user.unlink({cid: chal._id, keep});
+  });
+  await Q.all(challengesUserWasRemovedFrom);
 
-      if (leader === user._id || group.members.indexOf(leader) === -1) {
-        let seniorMember = _.find(group.members, m => m !== user._id);
+  let promises = [];
 
-        // could be missing in case of public guild (that can have 0 members) with 1 member who is leaving
-        if (seniorMember) update.$set = {leader: seniorMember};
-      }
+  // If user is the last one in group and group is private, delete it
+  if (group.memberCount === 1 && (
+    group.type === 'party' ||
+    group.type === 'guild' && group.privacy === 'private'
+  )) {
+    return await group.remove();
+  }
 
-      update.$inc = {memberCount: -1};
-      return Q.all([
-        model.update({_id: group._id}, update).exec(), // eslint-disable-line no-use-before-define
-        user.save(),
-      ]);
-    })(),
-  ]).then(() => {
-    firebase.removeUserFromGroup(group._id, user._id);
-    return; // TODO ok not to return promise?
-  }).catch(err => { // TODO do we have to catch err if we return the promise?
-    throw err;
-  });
+  // otherwise just remove a member TODO create User.methods.removeFromGroup?
+  if (group.type === 'guild') {
+    promises.push(User.update({_id: user._id}, {$pull: {guilds: group._id } }).exec());
+  } else {
+    promises.push(User.update({_id: user._id}, {$set: {'party._id': undefined } }).exec());
```

Comment:
Is there a fast way of doing this rather than setting all the values back to default inside the $set object?